superset/assets/javascripts/explore/AdhocMetric.js

```diff
@@ -1,7 +1,45 @@
+import { sqlaAutoGeneratedMetricRegex } from './constants';
+
+export const EXPRESSION_TYPES = {
+  SIMPLE: 'SIMPLE',
+  SQL: 'SQL',
+};
+
+function inferSqlExpressionColumn(adhocMetric) {
+  if (adhocMetric.sqlExpression && sqlaAutoGeneratedMetricRegex.test(adhocMetric.sqlExpression)) {
+    const indexFirstCloseParen = adhocMetric.sqlExpression.indexOf(')');
+    const indexPairedOpenParen =
+      adhocMetric.sqlExpression.substring(0, indexFirstCloseParen).lastIndexOf('(');
+    if (indexFirstCloseParen > 0 && indexPairedOpenParen > 0) {
+      return adhocMetric.sqlExpression.substring(indexPairedOpenParen + 1, indexFirstCloseParen);
+    }
+  }
+  return null;
+}
+
+function inferSqlExpressionAggregate(adhocMetric) {
+  if (adhocMetric.sqlExpression && sqlaAutoGeneratedMetricRegex.test(adhocMetric.sqlExpression)) {
+    const indexFirstOpenParen = adhocMetric.sqlExpression.indexOf('(');
+    if (indexFirstOpenParen > 0) {
+      return adhocMetric.sqlExpression.substring(0, indexFirstOpenParen);
+    }
+  }
+  return null;
+}
+
 export default class AdhocMetric {
   constructor(adhocMetric) {
-    this.column = adhocMetric.column;
-    this.aggregate = adhocMetric.aggregate;
+    this.expressionType = adhocMetric.expressionType || EXPRESSION_TYPES.SIMPLE;
+    if (this.expressionType === EXPRESSION_TYPES.SIMPLE) {
+      const inferredColumn = inferSqlExpressionColumn(adhocMetric);
+      this.column = adhocMetric.column || (inferredColumn && { column_name: inferredColumn });
```

Comment:
Just to make sure I understand, `inferSqlExpressionColumn` is necessary if adhocMetric was a sql expression typed previously when EXPRESSION_TYPES was SQL correct? There's a lot going on in this section, maybe we could add a few comments to clarify?